### PR TITLE
Add RuboCop to project

### PR DESCRIPTION
**Pull request by aried3r**
Hey!

This is just a proposal in the form of a PR. Instead of opening an issue I thought I'd go ahead and add a minimal working version. :)

I was wondering if you'd be ok to add a linter (in this case RuboCop) to this repo. While submitting my last few PRs I noticed that the code is lacking a uniform coding standard and is also contains some inconsistent whitespaces (sometimes trailing whitespaces or extra newlines etc).

While obviously this is very subjective, I just wanted to submit this as a mere suggestion, because I like this project a lot and would love to improve on it, even if it's mainly formatting.

Currently this deactivates all cops and puts them into the `.rubocop_todo.yml` file (done by `rubocop --auto-gen-config`). 

If this is a welcome change, I'd then go on to discuss which linters are wanted and little by little apply the necessary changes in separate PRs to avoid conflicts. For example, one of the major offenses (by count) is `Layout/IndentationWidth`, because this project uses 4 spaces to indent instead of 2 which is somewhat usual for other Ruby code (highly subjective of course). Here some decisions have to be made, follow the 4 spaces or apply the changes to adhere to the [community Ruby style guide](https://github.com/rubocop-hq/ruby-style-guide#spaces-indentation).

Anyway, hope this is a welcome addition and if this is accepted I hope we can find some cops that we can apply to this codebase!

### CODE REVIEW

**Comment by aried3r** [comment]
I just thought about 2.0 and 2.2 again. Maybe it's not that useful to have RuboCop run on _every_ Ruby version. The result should be the same anyway.

Maybe I can figure something out using `ENV` variables or use [build stages](https://docs.travis-ci.com/user/build-stages/). Maybe I can find a repo that uses that.

**Comment by jcupitt** [comment]
Maybe have two stages? Run rspec in the first over the matrix, then run rubocop in the second just on 2.4 or whatever.

Something like:

https://docs.travis-ci.com/user/build-stages/matrix-expansion/

**Comment by jcupitt** [comment]
Oh, fails because bundle will try to install rubocop in 2.0 

https://travis-ci.org/libvips/ruby-vips/jobs/436732008

I supposed we'd have to remove rubocop from the spec and install explicitly in the rubocop stage.

**Comment by aried3r** [comment]
The current test run works on all Rubies again and shows how RuboCop fails the build, because I actually added a RuboCop offense while coding that. Thing is, if I'd put it on one line it would exceed 100 characters per line, another thing that we should define, the `MaxLineWidth`.

File: Rakefile [hunk 1]
```diff
@@ -27,3 +27,9 @@ require "yard/rake/yardoc_task"
 
 YARD::Rake::YardocTask.new do |yard|
 end
+
+# RuboCop requires Ruby >= 2.2
+if Gem::Version.new(RUBY_VERSION) >= Gem::Version.new("2.2.0")
```
Comment (aried3r) [inline Rakefile]: Just saw I use `Gem.ruby_version` and `Gem::Version.new(RUBY_VERSION)` instead of using one consistently. Which version do you prefer? `Gem.ruby_version` already returns a `Gem::Version` which is nice.

**Comment by ioquatix** [comment]
Mojave has 2.3.7:

```
> /usr/bin/ruby --version
ruby 2.3.7p456 (2018-03-28 revision 63024) [universal.x86_64-darwin18]
```

**Comment by ioquatix** [comment]
So, you asked my opinion about RuboCop.

I don't use it, generally.

I can see value in it. For some use cases.

Personally, I find it unnecessary, and it also consumes time fixing things which are, practically speaking, irrelevant. My code is fine how it is. People have told me it's easy to read, well structured, etc. The only negative feedback I've ever had is that I prefer tabs rather than spaces for indentation, and mostly from people who are unable make any actual tangible contributions.

On the flip side, I've seen projects and PRs stunted by RuboCop. I've never seen a project that was easier to contribute to because of RuboCop. If anything, it's the things RuboCop can't fix that are actually more problematic.

So, basically, I'm slightly against it, and I think there are more important issues to address.

I think what would motivate me to support this PR, is if this PR actually makes @aried3r more likely to contribute tangible improvements in the future. Personally, I'd like to see an existing track record from a contributor, then make a suggestion like "My workflow would be easier of we had RuboCop", and then I'd be fully behind that because obviously that supports someone who is making a significant effort to better the project.

**Comment by ioquatix** [comment]
Some other less abstract points:

- Reformatting an entire code base breaks `git blame`
- It also makes it hard for forks to follow, e.g. my fork of `ruby-vips` would be tangibly affected by this.

File: Rakefile (hunk 1 above)
Comment (jcupitt) [reply Rakefile]: Oh not sure. I guess `Gem.ruby_version` is simpler.

**Comment by jcupitt** [comment]
For `MaxLineWidth`, I use 80 everywhere myself. ruby-vips is all 80 char lines now.

**Comment by jcupitt** [comment]
Thank you @ioquatix !

Yes, I agree reformatting would cause difficulties for git blame and for forks. There's `-w` to make it less painful, I suppose.

I've used layout checkers on all the other bindings and it hasn't been more than a slight annoyance, so far at least. libvips itself does not use one and that actually has been a problem -- libvips is in a roughly K&R style which I'm not sure I could enforce automatically and which no one but me likes. I end up having to reformat almost all PRs by hand, which is tedious.

Now ruby-vips is officially supported in Rails / Shrine / ImageProcessing, it feels like it is growing up a bit. A clear layout policy plus enforcement seems like good practice.

**Comment by aried3r** [comment]
Hey @ioquatix, thanks for your input!

**tl;dr:** I agree with most of your points, but still think it's worth doing and enabling a selected few cops in the beginning that are concerned with style (whitespace, line length) and then if wanted, enforce other more subjective cops, like method length, guard clauses or dangling commas in multiline statements.

**Long version:**

> I can see value in it. For some use cases.
> 
> Personally, I find it unnecessary, and it also consumes time fixing things which are, practically speaking, irrelevant.

I agree! At least in part. I think for formatting it's very useful for contributions, there's no need to look at other parts of the code on how to layout something, because a tool can help you with that. The easiest example being the line length.
Personally I would follow the style [paper_trail is using](https://github.com/paper-trail-gem/paper_trail/blob/4ea221d5478bddc7dbdf4955bebe9b74199ef7a4/.rubocop.yml#L45-L73), disabling most `Metrics` cops. AbcSize, method length, module length etc are – IMHO – very difficult to objectively quantify for the general case. A method with 100 lines might be too much, but that will probably be caught during code review, while 20 might be too little and annoying once you hit 21 lines.

So I think as a formatter, RuboCop can be very useful and also as a linter, but some other parts might be annoying and can be disabled.

Talking about `paper_trail`, its maintainer slowly introduced more and more cops and prefixed almost all of the commits with [`Lint: <name of cop>`](https://github.com/paper-trail-gem/paper_trail/search?q=Lint&type=Commits) which I think is nice.

> On the flip side, I've seen projects and PRs stunted by RuboCop. I've never seen a project that was easier to contribute to because of RuboCop.

Well, ideally RuboCop won't get in your way and you won't notice it, but RuboCop certainly isn't about making it easier to contribute, but easier to follow the style of the code base and spare maintainers commenting on PRs that a line has too many characters, there's RuboCop and GitHub bots for that.

> If anything, it's the things RuboCop can't fix that are actually more problematic.

I'm not sure which things you have in mind, but yeah, RuboCop won't write the code, nor will it find severe problems with it. But again, I think that's missing the point of having a consistent code style.

In this PR for example, all cops are disabled by default and need to be enabled and fixed one by one, which should ease the transition.

> I think what would motivate me to support this PR, is if this PR actually makes @aried3r more likely to contribute tangible improvements in the future. Personally, I'd like to see an existing track record from a contributor, then make a suggestion like "My workflow would be easier if we used RuboCop"

Well, let me be the first to say _"My workflow would be easier if we used RuboCop"_, which is why I opened this PR :)
When I [contributed](https://github.com/libvips/ruby-vips/pull/170) to this project (although with _a lot_ of help from @jcupitt), I was hindered by the current code style because there's so much trailing whitespace, that I had to configure all my editors to **not** remove them, **not** remove extraneous newlines at the end etc just for this project, and then still had to `git add -p` because I didn't immediately find some options that deal with extra whitespace.

> * Reformatting an entire code base breaks `git blame`

Yep, absolutely! I made that argument on a codebase I worked on before, but nowadays I don't really worry about it. When is the right time to introduce formatting? Right at the beginning would be perfect, sure. But is it then not allowed because our tooling (`git blame`) can't deal with it so well? The code is for humans to read and the the machine to execute. The machine doesn't care about formatting, but we maybe do. 
Your own code (which I appreciate a lot!) is a good example. You don't need RuboCop, because you adhere to a very strict style yourself. But for others, that might not be so clear and easy, RuboCop can help in an automated fashion to get feedback very fast instead of when maintainers have time to review.

Side note, `git blame`, even with `-w` or with tools like [`hyper-blame`](https://commondatastorage.googleapis.com/chrome-infra-docs/flat/depot_tools/docs/html/git-hyper-blame.html) aren't really sufficient for me in the console, where I use git the most. For blame I actually use GitHub these days:

kapture 2018-10-04 at 15 20 57

As for the forks, are there any forks being "maintained forks" that would really be affected negatively by this change in a way that cannot be resolved by some clever rebasing and cherry-picking?

**Comment by aried3r** [comment]
> I end up having to reformat almost all PRs by hand, which is tedious.

Yup, that's exactly a problem RuboCop can solve. Some projects use [HoundCI](https://houndci.com/) which can comment automatically on style violations. Sometimes it can be annoying, but as you said, better than having a human do that or doing it manually. See also https://robots.thoughtbot.com/introducing-hound

I'm not advocating to use Hound, I'm just saying that it can be useful to automate some checking of style and convention. :)

> For `MaxLineWidth`, I use 80 everywhere myself. ruby-vips is all 80 char lines now.

Great! Let's do that! 80 is actually the default by RuboCop, do you want me to explicitly define it in `.rubocop.yml` anyway? I grew to be a fan of explicitly setting these rules instead of having to look up defaults, but I'm fine either way.

**Comment by aried3r** [comment]
> Let's do that! 80 is actually the default by RuboCop, do you want me to explicitly define it in `.rubocop.yml` anyway?

Nevermind that, you actually have to set it manually so other cops don't recommend merging multiple lines into one if it exceeds the line length limit.

For example:

```Ruby
if very_long_conditional && another
  imagine_this_line_is_very_long_but_just_below_80_characters
end
```
 
Rubocop would suggest to use 
```Ruby
imagine_this_line_is_very_long_but_just_below_80_characters if very_long_conditional && another
```

which puts it above 80 chars. If you set LineLength to 80, this won't happen. I'll add this in the next commit.

**Comment by ioquatix** [comment]
I think those are all good points and as long as it's actually a net benefit I don't see any problem with it.

**Comment by aried3r** [comment]
Sounds good! :)

I'll actually remove the current LineLength: Max config for now so there will be no errors since there are indeed 97 lines that are above 80 characters.

I'd then start by reformatting some parts and either make individual PRs or one PR with multiple commits which we could then discuss.

**Comment by aried3r** [comment]
Having said that, maybe we can merge the two PRs by @janko-m before I go on applying RuboCop formatting?

**Comment by jcupitt** [comment]
Oh no! I'd forgotten about those PRs :-( what an idiot. 

They are merged now, and I updated the docs for 8.7.

**Comment by jcupitt** [comment]
Huh some en-dashes, oh dear. I've fixed them, if you rebase.

**Comment by jcupitt** [comment]
Or we could merge to master and slowly fix or ignore the 2400 infractions in a series of PRs.

**Comment by aried3r** [comment]
You wanna keep the en dashes? I'm fine either way, I can also regenerate the config once more and we can set it to allow non-ASCII comments.

**Comment by jcupitt** [comment]
The en-dashes are weird, I've removed them in master.

**Comment by aried3r** [comment]
Rebased, all green now

**Comment by jcupitt** [comment]
So ... all OK to merge, do you think?

**Comment by jcupitt** [comment]
OK, let's do it. Thanks for pushing this nice improvement through.
